routers/api/v1/user/app.go

```diff
@@ -231,3 +231,89 @@ func DeleteOauth2Application(ctx *context.APIContext) {
 
 	ctx.Status(http.StatusNoContent)
 }
+
+// GetOauth2Application get OAuth2 Application
+func GetOauth2Application(ctx *context.APIContext) {
+	// swagger:operation GET /user/applications/oauth2/{id} user userGetOAuth2Application
+	// ---
+	// summary: get an OAuth2 Application
+	// produces:
+	// - application/json
+	// parameters:
+	// - name: id
+	//   in: path
+	//   description: Application ID to be found
+	//   type: integer
+	//   format: int64
+	//   required: true
+	// responses:
+	//   "200":
+	//     "$ref": "#/responses/OAuth2Application"
+	appID := ctx.ParamsInt64(":id")
+	app, err := models.GetOAuth2ApplicationByID(appID)
+	if err != nil {
+		if models.IsErrOauthClientIDInvalid(err) || models.IsErrOAuthApplicationNotFound(err) {
+			ctx.NotFound()
+		} else {
+			ctx.Error(http.StatusInternalServerError, "GetOauth2ApplicationByID", err)
+		}
+		return
+	}
+
+	app.ClientSecret = ""
+
+	ctx.JSON(http.StatusOK, convert.ToOAuth2Application(app))
+}
+
+// UpdateOauth2Application update OAuth2 Application
+func UpdateOauth2Application(ctx *context.APIContext, data api.CreateOAuth2ApplicationOptions) {
+	// swagger:operation PATCH /user/applications/oauth2/{id} user userUpdateOAuth2Application
+	// ---
+	// summary: update an OAuth2 Application, this includes regenerating the client secret
+	// produces:
+	// - application/json
+	// parameters:
+	// - name: id
+	//   in: path
+	//   description: application to be updated
+	//   type: integer
+	//   format: int64
+	//   required: true
+	// - name: body
+	//   in: body
+	//   required: true
+	//   schema:
+	//     "$ref": "#/definitions/CreateOAuth2ApplicationOptions"
+	// responses:
+	//   "200":
+	//     "$ref": "#/responses/OAuth2Application"
+	appID := ctx.ParamsInt64(":id")
+
+	err := models.UpdateOAuth2Application(models.UpdateOAuth2ApplicationOptions{
+		Name:         data.Name,
+		UserID:       ctx.User.ID,
+		ID:           appID,
+		RedirectURIs: data.RedirectURIs,
+	})
+	if err != nil {
+		ctx.Error(http.StatusBadRequest, "", "error updating oauth2 application")
+		return
+	}
+	app, err := models.GetOAuth2ApplicationByID(appID)
```

Comment:
I would move **GetOAuth2ApplicationByID** bevore ***UpdateOAuth2Application**

Reply:
I'm concerned we would then return out of data information back to the client

Reply:
error handling is not the best ... we need a refactor there :(  ... I'm ok with this - but there should be a check if OAuth2App exist before Update :(

resolfe this for now. I'll send a refactor afterwards

Reply:
I think the proper fix here is to refactor the model to throw explicit errors, and to return the updated app, I'm okay with fixing it here, however I feel that it is out of scope of this PR.

Reply:
@dmolik leafe it like that - I'll take care if it later